framework/src/utils/SolutionInvalidity.C

```diff
@@ -0,0 +1,188 @@
+#include "MooseError.h"
+#include "MooseApp.h"
+#include "VariadicTable.h"
+
+// System Includes
+#include <chrono>
+#include <memory>
+#include <timpi/parallel_sync.h>
+
+// libMesh Includes
+#include "libmesh/parallel_algebra.h"
+#include "libmesh/parallel_sync.h"
+
+SolutionInvalidity::SolutionInvalidity(MooseApp & app)
+  : ConsoleStreamInterface(app),
+    ParallelObject(app.comm()),
+    _solution_invalidity_registry(moose::internal::getSolutionInvalidityRegistry())
+{
+}
+
+void
+SolutionInvalidity::flagInvalidSolutionInternal(InvalidSolutionID _invalid_solution_id)
+{
+  std::lock_guard<std::mutex> lock_id(_invalid_mutex);
+  if (_counts.size() <= _invalid_solution_id)
+    _counts.resize(_invalid_solution_id + 1);
+
+  ++_counts[_invalid_solution_id].counts;
+}
+
+bool
+SolutionInvalidity::solutionInvalid() const
+{
+  libmesh_parallel_only(comm());
+
+  bool is_invalid = false;
+  for (auto & entry : _counts)
+  {
+    if (entry.counts)
+    {
+      is_invalid = true;
+      break;
+    }
+  }
+
+  comm().max(is_invalid);
+  return is_invalid > 0;
+}
+
+void
+SolutionInvalidity::resetSolutionInvalid()
+{
+  for (auto & entry : _counts)
+    entry.counts = 0;
+}
+
+void
+SolutionInvalidity::resetSolutionInvalidTimeIter()
+{
+  for (auto & entry : _counts)
+    entry.timeiter_counts = 0;
+}
+
+void
+SolutionInvalidity::solutionInvalidAccumulation()
+{
+  for (auto & entry : _counts)
+  {
+    entry.total_counts += entry.counts;
+    entry.timeiter_counts += entry.counts;
+  }
+}
+
+void
+SolutionInvalidity::solutionInvalidAccumulationTimeIter()
+{
+  for (auto & entry : _counts)
+  {
+    entry.timeiter_counts += entry.counts;
+  }
+}
+
+void
+SolutionInvalidity::print(const ConsoleStream & console) const
+{
+  console << "\nSolution Invalid Warnings:\n";
+  summaryTable().print(console);
+}
+
+void
+SolutionInvalidity::sync()
+{
+  std::map<processor_id_type,
+           std::vector<std::tuple<std::string, std::string, unsigned int, unsigned int>>>
+      data_to_send;
+  if (processor_id() != 0)
+    for (const auto id : index_range(_counts))
+    {
+      const auto & entry = _counts[id];
+      if (entry.counts)
+      {
+        const auto & info = _solution_invalidity_registry.item(id);
+        data_to_send[0].emplace_back(
+            info.object_type, info.message, entry.counts, entry.timeiter_counts);
+      }
+    }
+
+  const auto receive_data = [this](const processor_id_type libmesh_dbg_var(pid), const auto & data)
+  {
+    mooseAssert(pid != 0, "Should not be used except processor 0");
+
+    for (const auto & [object_type, message, counts, timeiter_counts] : data)
+    {
+      InvalidSolutionID masterId = 0;
+      const moose::internal::SolutionInvalidityName name(object_type, message);
+      if (_solution_invalidity_registry.keyExists(name))
+        masterId = _solution_invalidity_registry.id(name);
+      else
+        masterId = moose::internal::getSolutionInvalidityRegistry().registerInvalidity(object_type,
```

Comment:
One last thing - this line isn't hit in coverage. We can talk today about how to hit it.

We need to have an invalid trigger happen on just one processor in a parallel run.

It's very important that we know this for sure works.

Reply:
also please remove master. Just call it mainId

Reply:
This comment has been addressed.